addons/point_of_sale/controllers/main.py

```diff
@@ -84,3 +85,132 @@ def print_sale_details(self, date_start=False, date_stop=False, **kw):
+                partner_values.update({'zip': partner_values.pop('zipcode', '')})
+                # Try to either find a partner based on the values, or create a new one if it's not possible.
+                pos_order.partner_id = self._get_matching_partner(partner_values, pos_order.company_id)
+                # Get the required fields for the invoice and add them to the context as default values.
+                with_context = {}
+                for field in additional_invoice_fields:
+                    with_context.update({f'default_{field.name}': invoice_values.get(field.name)})
+                # Allowing default values for moves is important for some localizations that would need specific fields to be set on the invoice, such as Mexico.
+                pos_order.with_context(with_context).action_pos_order_invoice()
+                return request.redirect('/my/invoices/%s?access_token=%s' % (pos_order.account_move.id, pos_order.account_move._portal_ensure_token()))
+
+        # Most of the time, the country of the customer will be the same as the order. We can prefill it by default with the country of the company.
+        if 'country_id' not in form_values:
+            form_values['country_id'] = pos_order.company_id.account_fiscal_country_id.id
+
+        return request.render("point_of_sale.ticket_validation_screen", {
+            'partner': request.env['res.partner'],
+            'format_amount': format_amount,
+            'env': request.env,
+            'countries': request.env['res.country'].sudo().search([]),
+            'states': request.env['res.country.state'].sudo().search([]),
+            'partner_can_edit_vat': True,
+            'pos_order': pos_order,
+            'invoice_required_fields': additional_invoice_fields,
+            'partner_required_fields': additional_partner_fields,
+            'access_token': access_token,
+            **form_values,
+        })
+
+    def _get_matching_partner(self, partner_vals, company_id):
+        """ Try to find a partner based on a multiple parameters.
+        - If the vat is not provided, try to match on the name and email.
+            - If a record match, use it, else create a new one.
+        - If it is provided, first match on it.
+            - If only one record match, use it.
+            - If no record match, create a new one.
+            - If there is multiple matches, try to match on TVA + name.
+                - If there is a match, reuse the partner and if not, create a new record.
+        """
+        ResPartner = request.env['res.partner']
+        if not partner_vals:
+            return ResPartner
+
+        if not partner_vals.get('vat'):
+            partner = ResPartner.sudo().search([
+                "|", ("company_id", "=", False), ("company_id", "=", company_id.id),
+                ("name", "=ilike", partner_vals.get('name')),
+                ("email", "=ilike", partner_vals.get('email'))
```

Comment:
at this point, both the name and email *must* be provided right? otherwise it must crash, not search for `None` instead

```suggestion
                "|", ("company_id", "=", False), ("company_id", "=", company_id.id),
                ("name", "=ilike", partner_vals['name']),
                ("email", "=ilike", partner_vals['email'])
```

(you could also see the criterion on `name` as unnecessary and duplicate-prone, but that's up to you)